src/core/ruby_version.rs

```diff
@@ -1003,81 +1011,291 @@ macro_rules! get_stack_frame_2_5_0(
+                            } else {
+                                log::debug!("TODO: Immediate object case is unhandled!");
+                            }
+                        }
+                    }
                 }
             }
 
-            let imemo: rb_method_entry_struct = source.copy_struct(env_me_cref).context(env_me_cref)?;
-            let klass: RClass_and_rb_classext_t = source.copy_struct(imemo.defined_class).context(imemo.defined_class)?;
-            if klass.classext.classpath == 0 {
-                return Err(anyhow::anyhow!("classpath was empty"));
+            if klass == 0 {
+                return Err(anyhow::anyhow!("klass was empty"));
+            }
+
+            let class_path = rb_class_path(klass, source)?;
+            Ok((class_path, singleton))
+        }
+
+        // https://github.com/ruby/ruby/blob/v3_3_0/vm_backtrace.c#L1841
+        pub fn profile_frame_full_label(
+            class_path: &str,
+            label: &str,
+            base_label: &str,
+            method_name: &str,
+            singleton: bool,
+        ) -> String {
+            let qualified = qualified_method_name(class_path, method_name, singleton);
+
+            if qualified.is_empty() || qualified == base_label.to_string() {
+                return label.to_string();
+            }
+
+            let label_length = label.len();
+            let base_label_length = base_label.len();
+            let mut prefix_len = label_length.saturating_sub(base_label_length);
+
+            // Ensure prefix_len doesn't exceed label length (defensive programming)
+            // Note: saturating_sub above already handles the < 0 case
+            if prefix_len > label_length {
+                prefix_len = label_length;
+            }
+
+            let profile_label = format!("{}{}", &label[..prefix_len], qualified);
+
+            if profile_label.is_empty() {
+                return String::new();
             }
-            let class_name = get_ruby_string(klass.classext.classpath as usize, source)?;
-            Ok(class_name.to_string())
+
+            // Get the prefix from label and concatenate with qualified_method_name
+            profile_label
         }
+    )
+);
 
+macro_rules! get_stack_frame_3_3_0(
+    () => (
         fn get_stack_frame<T>(
             iseq_struct: &rb_iseq_struct,
             cfp: &rb_control_frame_t,
             source: &T,
         ) -> Result<StackFrame> where T: ProcessMemory {
+            let cme = locate_method_entry(&cfp.ep, source)?;
+            let mut class_path = "".to_string();
+            let mut singleton = false;
+            let iseq = if cme != 0 {
+                let imemo: rb_method_entry_struct = source.copy_struct(cme).context(cme)?;
+                let method_type = source.copy(imemo.def as usize, 1).context(imemo.def as usize)?;
+                // https://github.com/ruby/ruby/blob/v3_3_0/internal/imemo.h#L21
+                let method_type = method_type[0] & 0xf;
```

Comment:
Note 0xf is because it is a bitfield of 4 bits, so we read 1 byte and then mask out the upper bits.